fax/implicit/twophase.py

```diff
@@ -32,98 +29,44 @@ def convergence_test(x_new, x_old):
         )
 
         return sol
-    return _default_solver
+    return _default_solve
 
 
-def make_adjoint_fixed_point_iteration(
-        param_func, default_rtol=1e-4, default_atol=1e-4, default_max_iter=5000,
-        default_batched_iter_size=1):
+@functools.partial(jax.custom_vjp, nondiff_argnums=(0, 1, 3))
+def two_phase_solve(param_func, init_xs, params, solvers=()):
```

Comment:
Thanks for pointing that out! I had misunderstood this issue and somehow convinced myself that it wasn't a problem here. I'll have to think about how best to handle this. Argument 1 can definitely be made non-static, in fact, I wouldn't want it to be static.

Argument 0 should be static but the hashing issues you raise are indeed problematic. I'll have to dig into this and make sure I understand better how `custom_vjp` handles `nondiff_argnums` since it might mean I should reconsider some parts of the API. I would expect `jax` would complain when verifying the shape/structure of whatever is returned if argument 0 wasn't static. Assume that this is the case, I really only see two options: 1) we revert to a `make_solver(...)` kind of api using closure for `param_func`, 2) we tolerate these limitations and document the caveats very clearly.

Reply:
When you're done your implementation, let me know, and I'll write some tests to check some of these leaked tracers.

Regarding the first argument, you're welcome to just lift [my code](https://github.com/NeilGirdhar/fax_suggestions/blob/master/cmm/fixed_point/combinator.py#L112).  (My use of `nondiff_argnums` is different because this uses my `custom_vjp` shim.)  Basically, I made all the arguments non-static, but since they're registered as JAX types, they can have static components.

I also stole some ideas from @manuel-delverme regarding [debuging helpers](https://github.com/NeilGirdhar/fax_suggestions/blob/master/cmm/fixed_point/iterated_function.py#L106).

Reply:
We'll delay solving this until we either switch to using `tjax` or decide to explore alternative solutions (#29).